src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs

```diff
@@ -0,0 +1,119 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Base.Deployments;
+using Stratis.Bitcoin.Features.Consensus.CoinViews;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Consensus
+{
+    /// <summary>Fetches coins for blocks that are likely to be validated in the near future to speedup full validation.</summary>
+    public class CoinviewPrefetcher : IDisposable
+    {
+        /// <summary>
+        /// How many blocks ahead prefetching will look.
+        /// When header at height X is dequeued block with header at height <c>X + Lookahead</c> will be prefetched in case block data is downloaded.
+        /// </summary>
+        /// TODO maybe make it dynamic so the value is increased when we are too close to the tip after prefetching was completed
+        /// and decreased if we are too far from the tip.
+        private const int Lookahead = 20;
+
+        /// <summary>Queue of headers that were added when block associated with such header was fully validated.</summary>
+        private readonly AsyncQueue<ChainedHeader> headersQueue;
+
+        private readonly ICoinView coinview;
+
+        private readonly CoinviewHelper coinviewHelper;
+
+        private readonly ConcurrentChain chain;
+
+        private readonly ILogger logger;
+
+        public CoinviewPrefetcher(ICoinView coinview, ConcurrentChain chain, ILoggerFactory loggerFactory)
+        {
+            this.coinview = coinview;
+            this.chain = chain;
+
+            this.headersQueue = new AsyncQueue<ChainedHeader>(this.OnHeaderEnqueuedAsync);
+            this.coinviewHelper = new CoinviewHelper();
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        /// <summary>
+        /// Prefetches UTXOs for a block and some blocks with higher height.
+        /// Prefetching is done on the background.
+        /// </summary>
+        /// <param name="header">Header of a block that is about to be partially validated and requires prefetching.</param>
+        private async Task OnHeaderEnqueuedAsync(ChainedHeader header, CancellationToken cancellation)
```

Comment:
Log header details - or will this be redundant when Fody interceptors are in place ?

Reply:
>Log header details - or will this be redundant when Fody interceptors are in place ?

Yes. That's why there are no entry\exit logs